app/Main.hs

```diff
@@ -229,25 +231,81 @@ getHttpManager opts = newManager managerSettings
+-- already existed or keep it (applying the inheritance blacklist if it exists).
+--
+-- Signals failure through a value of type VaultError.
+vaultEnv :: Context -> IO (Either VaultError [EnvVar])
 vaultEnv context = do
-  mountInfo <- requestMountInfo context
-  secrets <- mapExceptT (fmap $ first SecretFileError) $ readSecretList secretFile
-  secretEnv <- requestSecrets context mountInfo secrets
-  checkNoDuplicates (buildEnv secretEnv)
+  mountInfo <- doWithRetries getMountInfo
+  case mountInfo of
+    Left vaultError -> pure $ Left vaultError
+    Right mountInfo' -> do
+      secrets <- readSecretsFile secretFile
+      case secrets of
+        Left sfError -> pure $ Left $ SecretFileError sfError
+        Right secrets' -> do
+          -- NOTE: This retries fetching all the secrets if fetching one of the
+          -- secrets fails due to an error we should retry on. We would ideally
+          -- keep the secrets that were successfully fetched in memory and only
+          -- retry getting the ones we couldn't get yet, as this makes it less
+          -- likely to get a 'thundering herd' of Vaultenv processes all trying
+          -- and failing to get all of their secrets.
+          -- This would then need to happen in requestSecrets, but I don't want
+          -- to touch that right now.
+          -- Issue: https://github.com/channable/vaultenv/issues/90
+          secretEnv <- doWithRetries (getSecrets mountInfo' secrets')
+          case secretEnv of
+            Left vaultError -> pure $ Left vaultError
+            Right secretEnv' -> pure $ checkNoDuplicates (buildEnv secretEnv')
     where
+      doWithRetries :: (Retry.RetryStatus -> IO (Either VaultError a)) -> IO (Either VaultError a)
+      doWithRetries = Retry.retrying retryPolicy isRetryableFailure
+
+      -- | Indicator function for retrying to retry on VaultErrors (Lefts) that
+      -- shouldRetry thinks we should retry on. Needs to be in IO because the
+      -- actions to perform are in IO as well.
+      isRetryableFailure :: Retry.RetryStatus -> Either VaultError a -> IO Bool
+      isRetryableFailure _retryStatus (Right _) = pure False
+      isRetryableFailure retryStatus (Left err) = shouldRetry retryStatus err
+
+      retryPolicy = vaultRetryPolicy (cCliOptions context)
+
+      getMountInfo :: Retry.RetryStatus -> IO (Either VaultError MountInfo)
+      getMountInfo _retryStatus = catch (requestMountInfo context) httpErrorHandler
+
+      getSecrets :: MountInfo -> [Secret] -> Retry.RetryStatus -> IO (Either VaultError [EnvVar])
+      getSecrets mountInfo secrets _retryStatus = catch (requestSecrets context mountInfo secrets) httpErrorHandler
+
+      -- | "Handle" a HttpException by wrapping it in a Left VaultError.
+      -- We also edit the Request contained in the exception to remove the
+      -- Vault token, as it would otherwise get printed to stderr if we error
+      -- out.
+      httpErrorHandler (e :: HttpException) = case e of
```

Comment:
Please add a type signature for this binding.

Reply:
Done.